### PR TITLE
[rush] Differentiate remote and local execution in telemetry.

### LINKED ISSUE #4737: [rush] duplicated cobuild telemetry leading to data skew
## Summary

When using cobuilds, there is no easy way to determine if a given project was cobuilt or not during the `flushTelemetry` hook. You get a full list of operations and their execution time, both start + end time and nonCachedDurationMs. I opened #4680 as start + end time aren't super useful for cobuild cache hits. That work still causes data skew though. As a custom telemetry integrator, I'd like to make sure that I'm only counting operations once per cobuild for the agent that handled the building itself. No other agents should report on that operation.

#### Main Agent
```
"@company/my-package (test)": {
        "startTimestampMs": 31468.20680500008,
        "endTimestampMs": 40377.58414799906,
        "nonCachedDurationMs": 8649.619353000075,
        "result": "SUCCESS",
        "dependencies": [
          "@company/my-package (build)"
        ]
},
```

#### Build Cache Restore Agent
```
      "@company/my-package (test)": {
        "startTimestampMs": 220830.3333630003,
        "endTimestampMs": 220875.03935700096,
        "nonCachedDurationMs": 8649.619353000075,
        "result": "SUCCESS",
        "dependencies": [
          "@company/my-package (build)"
        ]
      },
```

For the primary agent, `endTimestampMs - startTimestampMs` doesn't match `nonCachedDurationMs`. On the build cache restore agent, there's no reliable way to determine if the package was built on this machine or another machine. We currently report both and have been accidentally introducing data skew to our metric collection.

### Recommendation

Possibly in conjunction with #4680, adding a new `wasCobuiltOnThisAgent` property to telemetry operation events would allow integrators to differentiate between primary and cache restore agents. I'd also recommend deprecating `nonCachedDurationMs` from telemetry, and moving to use just start and end time + `wasCobuildOnThisAgent`, as it's confusing to have multiple possible sources of truth for timing. It may be useful to capture build cache restore time elsewhere; for our use case though, it's not yet important to track.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| Would you consider contributing a PR? |  | Yes

### PR DESCRIPTION
**Pull request by aramissennyeydd**
## Summary

Fixes #4737. My goal is to address data skew questions before we go ahead with #4680 which just adjusts the data skew.

## Details

There is no great way currently to determine if telemetry for an operation was generated from the current machine or a remote machine. This is likely to cause data skew depending on how you ingest the Rush telemetry, either 
1. You restore duration from `nonCachedDurationMs`, which causes multiple events with the same duration (+/- a few milliseconds) if you emit events from each cobuild agent. That messes with averages and whatnot when aggregating your data.
2. You calculate duration from `startTimestampMs` and `endTimestampMs` which causes massive spikes in duration collected across your agents, as all but the primary agents report `0.05s` and the primary agent reports `15.00s`. That also messes with averages and whatnot during aggregation.

I propose a new `wasExecutedOnThisMachine` flag that monorepo maintainers can then use in their plugins to decide whether or not they want to process the given operation's data. 

## How it was tested

Tested in this repository, using the `sharded-repo` sandbox.

## Impacted documentation

Anything where Rush describes writing your own telemetry plugin.

### CODE REVIEW

**Review by iclanton — APPROVED** [review]

File: libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts [hunk 1]
```diff
@@ -845,10 +845,14 @@ export class PhasedScriptAction extends BaseScriptAction<IPhasedCommandConfig> {
...
 
           const { startTime, endTime } = operationResult.stopwatch;
+
           operationResults[operation.name!] = {
             startTimestampMs: startTime,
             endTimestampMs: endTime,
             nonCachedDurationMs: operationResult.nonCachedDurationMs,
+            wasExecutedOnThisMachine:
+              !operationResult.cobuildRunnerId ||
+              operationResult.cobuildRunnerId === cobuildConfiguration?.cobuildRunnerId,
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:855]: Strictly speaking, does a replay from cache count as "executed on this machine"?

Comment (aramissennyeydd) [reply libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:855]: Open to other names here. Strictly speaking, kind of/kind of not right? Cobuilds are intended to be state restores across machines as though it was built on any given cobuild agent, but from the data, that build wasn't executed on this machine b/c the cobuild runner ids from the state file and machine's cobuild config don't match.

Comment (aramissennyeydd) [reply libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:855]: @dmichon-msft Sorry lost track of this, would also be open to just adding the cobuildRunnerId to the metadata object and leaving it up to the user to make this check.

Comment (aramissennyeydd) [reply libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:855]: Or alternatively, we could leave as is and go forward with #4680, but advise folks to only send telemetry from a single agent.

**Comment by UberMouse** [comment]
bump on this would love to get it merged!

**Pull request merged**